pkg/iter/iterator.go

```diff
@@ -513,3 +513,54 @@ func ReadBatch(i EntryIterator, size uint32) (*logproto.QueryResponse, uint32, e
 	}
 	return &result, respSize, i.Error()
 }
+
+type entryWithLabels struct {
+	entry  logproto.Entry
+	labels string
+}
+
+type entryIteratorForward struct {
+	backwardIter      EntryIterator
+	cur               entryWithLabels
+	entriesWithLabels []entryWithLabels
+	loaded            bool
+	limit             uint32
+}
+
+// NewEntryIteratorBackward returns an iterator which loads all or upton N entries
+// of an existing iterator, and then iterates over them backward.
+func NewEntryIteratorForward(it EntryIterator, limit uint32) (EntryIterator, error) {
```

Comment:
To keep in mind for later we could make this a BufferedIterator where the limit would the the buffer size and keep rolling entry until the iterator is exhausted.